lang/funcs/collection.go

```diff
@@ -1052,50 +1062,71 @@ var SliceFunc = function.New(&function.Spec{
 		},
 	},
 	Type: func(args []cty.Value) (cty.Type, error) {
-		return args[0].Type(), nil
-	},
-	Impl: func(args []cty.Value, retType cty.Type) (ret cty.Value, err error) {
-		inputList := args[0]
-		if !inputList.IsWhollyKnown() {
-			return cty.UnknownVal(retType), nil
-		}
-		var startIndex, endIndex int
+		arg := args[0]
+		argTy := arg.Type()
 
-		if err = gocty.FromCtyValue(args[1], &startIndex); err != nil {
-			return cty.NilVal, fmt.Errorf("invalid start index: %s", err)
-		}
-		if err = gocty.FromCtyValue(args[2], &endIndex); err != nil {
-			return cty.NilVal, fmt.Errorf("invalid start index: %s", err)
+		if !argTy.IsListType() && !argTy.IsTupleType() {
+			return cty.NilType, fmt.Errorf("can only flatten lists, and tuples")
 		}
 
-		if startIndex < 0 {
-			return cty.NilVal, fmt.Errorf("from index must be >= 0")
-		}
-		if endIndex > inputList.LengthInt() {
-			return cty.NilVal, fmt.Errorf("to index must be <= length of the input list")
+		if argTy.IsListType() {
+			return argTy, nil
 		}
-		if startIndex > endIndex {
-			return cty.NilVal, fmt.Errorf("from index must be <= to index")
+
+		startIndex, endIndex, err := sliceIndexes(args, args[0].LengthInt())
+		if err != nil {
+			return cty.NilType, err
 		}
 
-		var outputList []cty.Value
+		return cty.Tuple(argTy.TupleElementTypes()[startIndex:endIndex]), nil
+	},
+	Impl: func(args []cty.Value, retType cty.Type) (ret cty.Value, err error) {
+		inputList := args[0]
 
-		i := 0
-		for it := inputList.ElementIterator(); it.Next(); {
-			_, v := it.Element()
-			if i >= startIndex && i < endIndex {
-				outputList = append(outputList, v)
-			}
-			i++
+		startIndex, endIndex, err := sliceIndexes(args, inputList.LengthInt())
+		if err != nil {
+			return cty.NilVal, err
 		}
 
-		if len(outputList) == 0 {
+		if endIndex-startIndex == 0 {
+			if retType.IsTupleType() {
+				return cty.EmptyTupleVal, nil
+			}
 			return cty.ListValEmpty(retType.ElementType()), nil
 		}
+
+		outputList := inputList.AsValueSlice()[startIndex:endIndex]
+
+		if retType.IsTupleType() {
+			return cty.TupleVal(outputList), nil
+		}
+
 		return cty.ListVal(outputList), nil
 	},
 })
 
+func sliceIndexes(args []cty.Value, max int) (int, int, error) {
+	var startIndex, endIndex int
+
+	if err := gocty.FromCtyValue(args[1], &startIndex); err != nil {
+		return 0, 0, fmt.Errorf("invalid start index: %s", err)
+	}
+	if err := gocty.FromCtyValue(args[2], &endIndex); err != nil {
+		return 0, 0, fmt.Errorf("invalid start index: %s", err)
+	}
+
+	if startIndex < 0 {
+		return 0, 0, fmt.Errorf("from index must be >= 0")
```

Comment:
This is totally subjective, but I think the `>=` and `<=` symbols in these messages make them read a little weirdly, particularly for `<=` which I initially interpreted as a left-facing arrow rather than "less than or equal to". Could we write this out in full words instead, to make sure the meaning is clear? (We have other error messages that are much longer than these would be with that expansion!)